### PR TITLE
Use context free expression types in evolving array checking and cache context free type

### LINKED ISSUE #26031: Intellisense and Go to Symbol not working with JavaScript
_From @israelnyc on July 26, 2018 3:8_

Issue Type: <b>Bug</b>

I've been experiencing this issue starting with release 1.24 where intellisense and Go to Symbol (CTRL+SHIFT+O) don't work at all, with certain JavaScript files. All I see is the Loading... message or the animated progress bar. 

At first I thought it may have been related to large JavaScript files and the amount of symbols they contain since I am working with files some of which have thousands of lines of code. However, after trying with different projects, that wasn't the case. 

I think I have narrowed it down though to a simplified reproduction, at least based on the code I'm experiencing the issue with. I've been able to consistently reproduce it with the sample code below on two separate Windows 10 systems as well as on macOS 10.13.3, all with the latest insiders build with extensions disabled or no extensions installed at all. 

It seems that when updating an array by doing `arr[arr.length] = 'value'` vs. `arr.push('value')` multiple times, it causes intellisense to slow down drastically or not work at all. With the code below it is either really slow and eventually works or just doesn't work at all. Performance seems to get worse with each added line. 

When it fails to work, the process CPU usage also stays between 10% and 15% when those lines are in the file. Whereas if they are commented out or when in version 1.23.1, the process stays below 1%. 

vscode-cpu-compare

With the actual code I'm working with, there are many more such lines updating arrays and I think that is what causes it to never work at all. The only time I have seen it work is if I try it quickly after first opening the editor and before the Outline for the file is loaded. But that is inconsistent and it won't work on subsequent attempts.

Something else I noticed with the code below is that the issue will randomly not occur - seemingly when the editor has been left open a while and hasn't been touched - but only for a short while or until the file is modified. 

It works perfectly fine and with zero delay on version 1.23.1, but every release since then on both the stable and insiders versions have the same issue. 

``` javascript
function build() {
    var arr = [];

    arr[arr.length] = 'value'; 
    arr[arr.length] = 'value'; 
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value'; 
    arr[arr.length] = 'value'; 
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value'; 
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value';
    arr[arr.length] = 'value'; 

    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
    // arr.push('value');
}
```

VS Code version: Code - Insiders 1.26.0-insider (5fae91d61f12726e149fe5baed342eac2fbf8f88, 2018-07-25T02:04:01.629Z)
OS version: Windows_NT x64 10.0.17134

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz (12 x 3696)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|31.93GB (20.87GB free)|
|Process Argv|C:\Users\Israel\AppData\Local\Programs\Microsoft VS Code Insiders\Code - Insiders.exe|
|Screen Reader|no|
|VM|0%|

</details>Extensions: none

_Copied from original issue: Microsoft/vscode#55104_

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #26031

Also includes another similar test case we found in the wild from https://github.com/archilogic-com/3dio-js which fails in a slightly different way prior to this fix because it is right-recursive in the control flow graph instead of left-recursive.

This fix is simply to use the context free expression type [added here](https://github.com/Microsoft/TypeScript/commit/5c2091ad33d3d0446fbfecd7d5efeeafb3ccfa10) for the LHS same as we do on the RHS and cache it in both scenarios, since it is context free. To ensure this is safe, I've also swapped it to using `CheckMode.SkipContextSensitive` (same as the other scenarios where we cache a context free type during inference to produce better inferences), to avoid reporting incorrect or partial types for partially solved context sensitive lambdas and function expressions (which would otherwise be a check order caching hazard). This caching keeps the evolving array type flow control analysis time linear, even when the expressions are both left and right recursive on their own reference.

### CODE REVIEW

File: tests/cases/compiler/deeplyDependentLargeArrayMutation2.ts [hunk 1]
```diff
@@ -0,0 +1,498 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @target: es6
+// @filename: foo.js
+// Exerpt from https://github.com/archilogic-com/3dio-js - MIT Licenced
```
Comment (billti) [inline tests/cases/compiler/deeplyDependentLargeArrayMutation2.ts]: *excerpt*

File: tests/baselines/reference/deeplyDependentLargeArrayMutation2.types [hunk 2]
```diff
@@ -0,0 +1,3606 @@
...
+>cvPos : number
+>13 : 13
+>nY : number
+
+            closetVertices[cvPos + 8] = closetVertices[cvPos + 14] = iZ;
+>closetVertices[cvPos + 8] = closetVertices[cvPos + 14] = iZ : any
+>closetVertices[cvPos + 8] : any
+>closetVertices : any[]
+>cvPos + 8 : number
+>cvPos : nu
```
Comment (billti) [inline tests/baselines/reference/deeplyDependentLargeArrayMutation2.types]: Do we not have a reduced repro we can baseline? This seems more appropriate for the RWC tests. I can't imagine anyone actually reviewing this to any depth.

**Comment by billti** [comment]
I see the Node8 run still failed with `RangeError: Maximum call stack size exceeded` on a test.

**Comment by weswigham** [comment]
The large control flow graph test, no less - and only on node 8 but not 6 or 10. 🤔  Iiiiiinteresting.

File: tests/baselines/reference/deeplyDependentLargeArrayMutation2.types (hunk 2 above)
Comment (weswigham) [reply tests/baselines/reference/deeplyDependentLargeArrayMutation2.types]: I simplified it, but it's not much shorter, just way less obvious what it's motivation is. To hopefully prevent someone brushing off a regression in it in the future, I've included a comment to the effect of why the test exists and what more obvious pattern it represents.

**Comment by weswigham** [comment]
Hmmm... theory: the new local within `getContextFreeTypeOfExpression` takes up enough space on the stack in node 8 to blow the stack (since in v8 stack size is apparently a MB limit, not a depth limit). I just successfully ran the test in node8 under the debugger just by adding a logpoint of the `flowDepth` stack variable after where the stack overflow usually occurs and then removing it (which throws the function into a deoptimized state which may mess with memory usage a bit), so it's certainly something wonky; feels like `node`'s not tracking `flowDepth` correctly or something.

In any case, lowing the maximum flow depth a tad seems to solve the problem - it probably allows a bit more stack space for the new variable.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -14407,8 +14407,8 @@ namespace ts {
             return resultType;
 
             function getTypeAtFlowNode(flow: FlowNode): FlowType {
-                if (flowDepth === 2500) {
-                    // We have made 2500 recursive invocations. To avoid overflowing the call stack we report an error
+                if (flowDepth === 2250) {
```
Comment (billti) [inline src/compiler/checker.ts]: Interesting (i.e. arbitrary :-p ) choice. Seems like if we know 2500 blows the stack in some tests, we might want a little more headroom.

What kind of patterns cause it to get this high? Are there completely valid patterns that require a number up in this range? (i.e. Say hypothetically we set it to 1000. Would there be a noticeably degradation in behavior for a meaningful set of users?)

Comment (weswigham) [reply src/compiler/checker.ts]: Deeply nested expressions and long statement chains or a combination thereof. Our own codebase gets to around 350 and I've seen some js codebases hit around 1000 before - we should probably try to keep it high, IMO.

**Pull request merged**